src/LibraryManager/CacheService/CacheService.cs

```diff
@@ -0,0 +1,128 @@
+        }
+
+        /// <summary>
+        /// Retruns the library content from the provider url and writes to cacheFile
+        /// </summary>
+        /// <param name="url"></param>
+        /// <param name="cacheFile"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task<string> GetLibraryAsync(string url, string cacheFile, CancellationToken cancellationToken)
+        {
+            return await GetResourceAsync(url, cacheFile, _libraryExpiresAfterDays, cancellationToken);
+        }
+
+        /// <summary>
+        /// Downloads a resource from specified url to a destination file
+        /// </summary>
+        /// <param name="url"></param>
+        /// <param name="fileName"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        private async Task DownloadToFileAsync(string url, string fileName, CancellationToken cancellationToken)
+        {
+            Stream libraryStream = await _requestHandler.GetStreamAsync(url, cancellationToken);
+            await FileHelpers.WriteToFileAsync(fileName, libraryStream, cancellationToken);
+        }
+
+        private async Task<string> GetResourceAsync(string url, string localFile, int expiration, CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+
+            if (!File.Exists(localFile) || File.GetLastWriteTime(localFile) < DateTime.Now.AddDays(-expiration))
+            {
+                await DownloadToFileAsync(url, localFile, cancellationToken).ConfigureAwait(false);
+            }
+
+            return await FileHelpers.ReadFileTextAsync(localFile, cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Refreshes the cache for the given set of files if expired 
+        /// </summary>
+        /// <param name="librariesCacheMetadata"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task HydrateCacheAsync(IEnumerable<CacheServiceMetadata> librariesCacheMetadata, CancellationToken cancellationToken)
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+
+            List<Task> hydrateTasks = new List<Task>();
+
+            foreach (CacheServiceMetadata metadata in librariesCacheMetadata)
+            {
+                if (!File.Exists(metadata.DestinationPath) || File.GetLastWriteTime(metadata.DestinationPath) < DateTime.Now.AddDays(-_libraryExpiresAfterDays))
+                {
+                    Task readFileTask = DownloadToFileAsync(metadata.Source, metadata.DestinationPath, cancellationToken);
```

Comment:
Will this potentially lead to firing off many many network requests in parallel?

Reply:
Yes, we decided it was ok for files within libraries. We are still restoring libraries sequentially

Reply:
What is the threshold that is acceptable?  What if a library has 50 files (no joke)?